### ISSUE TITLE
[feature request] ability to remap Lightgun buttons

### ISSUE THREAD

**Issue by Widge-5** [libretro/pcsx_rearmed#624]
Currently, it is not possible to remap lightgun controls. In the Retroarch 'Main Menu | Settings | Input | Port Controls X' menus all the controller buttons can be reconfigured, but altering gun button mappings makes no difference to the buttons' behaviour in game. Gun buttons seem to be hardcoded into the core as follows - trigger=Mouse1, reload=Mouse3, AuxA=Mouse2 and despite changing these in the RA menus or retroarch.cfg, they cannot be changed in practice.
So there are 3 parts to this request:

1 : The ability to remap the gun buttons and swap them around as you can with control pads, or in other cores;
2 : The ability to also map keyboard presses to the gun buttons as you can with control pad mappings (in order to allow pedal controls to be used);
3 : The ability to map 2-button combinations to a single button on the gun (some games require both guncon buttons to be pressed simultaneously to act as a Start button - mapping this to a single button press would be a real feature).

Points 2&3 will not be possible until point 1 has been fulfilled, hence the shared request.

I am using lr-pcsx-rearmed with RetroPie on a Raspberry Pi 4B. My Retroarch version is 1.9.9.

(This is my first request, so I apologise if I haven't done it correctly).

**Comment by StormedBubbles** [comment]
I tinkered with the first two parts of the request by editing /frontend/libretro.c. I am having a little trouble understanding the code (I am an amateur, so bear with me 😜), but I was able to accomplish mapping the trigger to the Select button, the GunCon A button to the Start button, and the GunCon B button to the D-pad up button with these changes:

```
   // Trigger
   if (ret & (1 << RETRO_DEVICE_ID_LIGHTGUN_TRIGGER))
      in_keystate[port] |= (1 << DKEY_CIRCLE);

   // A
   if (ret & (1 << RETRO_DEVICE_ID_LIGHTGUN_AUX_A))
      in_keystate[port] |= (1 << DKEY_START);

   // B
   if (ret & (1 << RETRO_DEVICE_ID_LIGHTGUN_AUX_B))
      in_keystate[port] |= (1 << DKEY_CROSS);
```

The GunCon A part makes sense, but I don't understand why mapping the trigger to circle and the GunCon B button to cross ended up getting me select and up, respectively. The user can then map whatever mouse, keyboard, or controller buttons he/she wants to any of those in the RetroArch menu and have free rein over the GunCon buttons. The code for pcsx_rearmed includes an offscreen-shot button labeled as "trigger 2" in a few places, but I wasn't able to map that to anything. The emulator already allows the user to "shoot off screen" by reaching a certain threshold along the border of the viewable area, so I guess this one isn't as important as the others.

I hope that someone else can come along and make more proper RetroArch mappings of:

- Trigger = Gun Trigger
- GunCon A = Gun Aux A
- GunCon B = Gun Aux B
- Offscreen shot = Gun Reload
